docs/design/drf - fairshare.md

```diff
@@ -0,0 +1,179 @@
+
+DRF plugin use preemption and order of job to balance resource among jobs. The `share` in this plugin is defined as resource usage, the higher `share` means this job occupies the more resource now.
+
+#### Namespace Compare
+
+To introduce namespace weight into this plugin, we should define how to compare namespace having weight firstly.
+
+For namespace n1 having weight w1 and namespace n2 having weight w2, we can compute the `share` of resource and recorded as u1 and u2. Now, the resource usage of n1 less than n2 can be defined as (u1 / w1 < u2 / w2)
+
+`e.g.` ns1 having weight w1=2 use 6cpu, ns2 having weight w2=1 use 2cpu. In the scope of cpu, the ns1 use less resource than ns2. (6 / 3 < 3 / 1)
+
+#### Namespace Order
+
+Register `NamespaceOrder` function using the strategy mentioned above.
+
+#### preemption
+
+> The `preempt` action is disabled now. Do this later.
+
+In the `preemption` function now, strategy is just simply comparing the resource share among jobs .
+
+After adding namespace weight, we should check namespace of preemptor and preemptee firstly. The job in namespace with less resources can preempt others, or when namespace resource usage are the same, compare share of job instead.
+
+### Feature Interaction
+
+#### preempt action
+
+Preempt is a strategy set to choose victims and finally evict it.
+
+The way to choose victims is a function set named `Preemptable` registered by plugins. Such as job ready protection of Gang plugin, special pod protection of Conformance plugin, job share balance strategy of DRF plugin.
+
+All these plugin would choose some victims respective, and the intersection of them would be the final victim set. So, the choice made by DRF plugin would never break the requirement of others.
+
+### short hand
+
+1. Preempt may cause killing of some running pod.
+
+### Cases:
+
+- cluster have __16 cpu__, queue and namespace have default weight.
+
+    | queue | namespace | requested | queue assigned | namespace assigned |
+    | ----- | --------- | --------- | -------------- | ------------------ |
+    | q1    | ns1       | 5 cpu     | 8 cpu          | 4 cpu              |
+    |       | ns2       | 10 cpu    |                | 4 cpu              |
+    | q2    | ns3       | 10 cpu    | 8 cpu          | 6 cpu              |
+    |       | ns4       | 2 cpu     |                | 2 cpu              |
+
+- cluster have __16 cpu__, q1 with weight 1, q2 with weight 3. ns1 with weight 3, ns2 have weight 1, ns3 have weight 2, ns4 have weight 6.
+
+    | queue | namespace | requested | queue assigned | namespace assigned |
+    | ----- | --------- | --------- | -------------- | ------------------ |
+    | q1 w1 | ns1 w3    | 5 cpu     | 4 cpu          | 3 cpu              |
+    |       | ns2 w1    | 10 cpu    |                | 1 cpu              |
+    | q2 w3 | ns3 w2    | 10 cpu    | 12 cpu         | 10 cpu             |
+    |       | ns4 w6    | 2 cpu     |                | 2 cpu              |
```

Comment:
namespace4 gets more compared with namespace3 why?

Reply:
According to the design, ns3 should have only 3 * CPU and 9 * CPU for ns4, i think.